indexing-service/src/main/java/org/apache/druid/indexing/common/actions/LocalTaskActionClient.java

```diff
@@ -76,11 +76,34 @@ public <RetType> RetType submit(TaskAction<RetType> taskAction)
     }
 
     final long performStartTime = System.currentTimeMillis();
-    final RetType result = taskAction.perform(task, toolbox);
+    final RetType result;
+    if (isBatchAllocateAction(taskAction)) {
+      result = performBatchAllocateAction(taskAction);
+    } else {
+      result = taskAction.perform(task, toolbox);
+    }
     emitTimerMetric("task/action/run/time", taskAction, System.currentTimeMillis() - performStartTime);
     return result;
   }
 
+  private <R> boolean isBatchAllocateAction(TaskAction<R> taskAction)
+  {
+    return toolbox.canBatchSegmentAllocation()
+        && taskAction instanceof SegmentAllocateAction;
```

Comment:
could we instead add a method in TaskAction class itself? 
`shouldPerformAsync` and then call `performAsync` instead? 
The method can take toolbox as a parameter and thus the impl in `SegmentAllocationAction` can also check `toolbox.canBatchSegmentAllocation()`

Reply:
Add `canPerformAsync` and `performAsync` to `TaskAction`.